src/composition.rs

```diff
@@ -547,240 +474,155 @@ impl ToString for Composition {
+            .iter()
+            .position(|statement| {
+                matches!(statement, ast::Statement::Expr(_))
+            })
+            .unwrap_or_else(|| file.body.len());
+
+        file.body.insert(
+            statement_index,
+            ast::Statement::Expr(Box::new(ast::ExprStmt {
+                base: ast::BaseNode::default(),
+                expression: ast::Expression::PipeExpr(Box::new(
+                    analyzer.build(),
+                )),
+            })),
+        );
+
+        Self {
+            file,
+            analyzer,
+            statement_index,
         }
+    }
 
+    /// Sync the composition statement with the analyzer.
+    // This is, for obvious reasons, inefficient. It moves the items in the vec
+    // around a lot. Premature optimization and blah blah blah.
+    fn sync(&mut self) {
+        self.file.body.remove(self.statement_index);
         self.file.body.insert(
-            0,
+            self.statement_index,
             ast::Statement::Expr(Box::new(ast::ExprStmt {
                 base: ast::BaseNode::default(),
                 expression: ast::Expression::PipeExpr(Box::new(
-                    statement,
+                    self.analyzer.build(),
                 )),
             })),
         );
-        Ok(())
     }
 
-    pub(crate) fn add_measurement(
+    /// Resolve composition with a new ast File.
+    ///
+    /// This is the logic that re-attaches to the new composition. The logic
+    /// is complex, but the core of the work is the analyzer, which checks
+    /// all the statements to find the matching one.
+    ///
+    /// In the event the composition can't re-attach to the new AST, usually
+    /// through ambiguity, an error is returned. The Composition should then
+    /// be discarded.
+    pub fn resolve_with_ast(
         &mut self,
-        measurement: String,
-    ) -> CompositionResult {
-        let mut visitor =
-            CompositionStatementFinderVisitor::default();
-        flux::ast::walk::walk(
-            &mut visitor,
-            flux::ast::walk::Node::File(&self.file),
-        );
-        if visitor.statement.is_none() {
+        file: ast::File,
+    ) -> Result<(), ()> {
+        self.file = file;
+        let matches = self
+            .file
+            .body
+            .iter()
+            .enumerate()
+            .filter(|(_index, statement)| {
+                matches!(statement, ast::Statement::Expr(_))
+            })
+            .map(|(index, statement)| {
+                let expression = match statement {
+                    ast::Statement::Expr(expr) => expr,
+                    _ => unreachable!("Previous filter failed"),
+                };
+                let analyzer = CompositionStatementAnalyzer::analyze(
+                    *expression.clone(),
+                );
+                (index, analyzer)
+            })
+            .filter(|(_index, analyzer)| {
+                analyzer.clone() == self.analyzer
```

Comment:
The "is the same composition check" is based all the same composition bits being present. a.k.a the bucket, measurements, etc. What it does not check, is if extra transformations are in the middle.

Meaning, this will still resolve:
```
from(bucket: "myBucket")
|> range(start: v.timeRangeStart, stop: v.timeRangeStop)
|> sort()
|> filter(fn: (r) => r._measurement == "myMeasurement")
```

Then after a composition change is made, the resulting query will have moved any unrecognized transformations (`CompositionAnalyzer.calls`) to the bottom. For example, if we add a field then we get:
```
from(bucket: "myBucket")
|> range(start: v.timeRangeStart, stop: v.timeRangeStop)
|> filter(fn: (r) => r._measurement == "myMeasurement")
|> filter(fn: (r) => r._field == "newlyAddedFieldFilter")
|> sort()
```

I think this behavior is what we want (for now), but can we have a test that explicitly defines this use case? (Or did I miss that test?)

Reply:
I don't think we should add tests for things that we aren't asserting today. This implementation is incomplete and naive, but is definitely a step in the right direction, unblocking other work, and leaves space to continue massaging this code. Adding tests for unintended behavior is not something we should do.